### PR TITLE
Migrate Core Id tests from Groovy to Java

### PR DESCRIPTION
**Pull request by wborn**
Migrates the Core Id tests from Groovy to Java.

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Test
+    public void readFromPersistedFile() throws FileNotFoundException, IOException {
+        // we first need to remove the cached value
+        InstanceUUID.uuid = null;
+        File file = new File(ConfigConstants.getUserDataFolder() + File.separator + InstanceUUID.UUID_FILE_NAME);
+        file.getParentFile().mkdirs();
+        try (OutputStream os = new FileOutputStream(file)) {
+            IOUtils.write("123", os);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java]: As we are using Java 8 you can use `Files` and one of its `write` methods.
So we do not read to rely on `IOUtils` anymore.

See e.g. https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#write-java.nio.file.Path-java.lang.Iterable-java.nio.file.OpenOption...-

File: bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Test
+    public void ignoreEmptyFile() throws FileNotFoundException, IOException {
+        // we first need to remove the cached value
+        InstanceUUID.uuid = null;
+        File file = new File(ConfigConstants.getUserDataFolder() + File.separator + InstanceUUID.UUID_FILE_NAME);
+        file.getParentFile().mkdirs();
+        try (OutputStream os = new FileOutputStream(file)) {
+            IOUtils.write("", os);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java]: See above

File: bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java (hunk 1 above)
Comment (wborn) [reply bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java]: That's a good idea! It removes a dependency and further cleans up the code. :-)

Comment (wborn) [reply bundles/core/org.eclipse.smarthome.core.id.test/src/test/java/org/eclipse/smarthome/core/id/UUIDTest.java]: Looks like I've cleaned up to much now. ;-)

File: bundles/core/org.eclipse.smarthome.core.id.test/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -1,21 +1,16 @@
...
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-SymbolicName: org.eclipse.smarthome.core.id.test;singleton:=true
+Bundle-Vendor: Eclipse.org/SmartHome
+Bundle-Version: 0.10.0.qualifier
+Fragment-Host: org.eclipse.smarthome.core.id
+Import-Package: 
+ org.eclipse.jdt.annotation;resolution:=optional,
+ org.eclipse.smarthome.test,
+ org.eclipse.smarthome.test.storage,
+ org.hamcrest;core=split,
```
Comment (wborn) [inline bundles/core/org.eclipse.smarthome.core.id.test/META-INF/MANIFEST.MF]: Is using Hamcrest matchers still preferred or is its faith going to be the same as Groovy?
There's the trade off between readability and adding an extra dependency.

Comment (maggu2810) [reply bundles/core/org.eclipse.smarthome.core.id.test/META-INF/MANIFEST.MF]: Hm, why do we need them at all?
Can't we change `assertThat(uuid, is(equalTo("123")));` to `assertEquals("123", uuid)` (static import of Assert.* from JUnit)?
IMHO `assertEquals` is readable :wink: 
WDYT?

Comment (wborn) [reply bundles/core/org.eclipse.smarthome.core.id.test/META-INF/MANIFEST.MF]: Hamcrest may be overkill in a project like this with just a few assertions.

When projects grow and start to contain more and complexer tests, it's easier to immediately understand the intention of tests when you can read the Hamcrest assertThat statements like normal written sentences.

Though you can also use streams with predicates nowadays which make up for good alternatives for methods like anyOf/allOf in which Hamcrest really had an edge in the pre-Java-8 days.

Comment (maggu2810) [reply bundles/core/org.eclipse.smarthome.core.id.test/META-INF/MANIFEST.MF]: It seems in this test case there is no need to use Hamcrest at all.
Would you like to remove the Hamcrest usage for that test now or do you prefer to leave it in place?

Comment (wborn) [reply bundles/core/org.eclipse.smarthome.core.id.test/META-INF/MANIFEST.MF]: I'll remove it from this project since it's just a few asserts.

**Comment by maggu2810** [comment]
Thank you for your work.

**Pull request merged**